### ISSUE TITLE
CA-GMRES solver 

### ISSUE THREAD

**Issue by maddyscientist** [lattice/quda#394]
We need a solver for the bottom coarse grid that is robust and fast.  The presently used GCR is limited by the latency invoked by reductions.  A good candidate will be communication avoiding GMRES (CA-GMRES) since it robust and drastically reduced the synchronization versus the presently used GCR solver.

**Comment by alexstrel** [comment]
I was thinking about implementing deflated version of CA-GMRES in different context, namely to improve performance of GMRESDR; this is  reported ,e,g, here: www.icl.utk.edu/sites/icl/.../icl-utk-762-2014.pdf This solver does require all ingredients of a regular CA-GMRES , so our experience would be essential for deflation stuff as well. (Corresponding task for AMG would be a pure CPU implementation, though)

**Comment by maddyscientist** [comment]
I don't think it should make any difference for implementation whether this is CPU or GPU.  We have all the routines we need on both: BLAS and stencil can run on either; and MAGMA can deal with CPU data.  So I think it should not matter.

On another topic, I guess once your deflation branch is merged, this will need to be updated so that the solvers can run on CPU or GPU.  We'll need this to deflate the coarsest system.  I guess this should be easy though?

**Comment by alexstrel** [comment]
Absolutely, the solver itself relies on MAGMA, that can use either CPU or GPU (depending on the problem size).

**Comment by maddyscientist** [comment]
Just leaving a note for myself :wink: 

A further advantage of CA-GMRES (and communication-avoiding solvers in general) is that they allow for the introduction of the so-called matrix-powers kernel.  Such an approach is especially beneficial for the coarse grids: on the coarsest grids we don't typically saturate the memory bandwidth due to lack of parallelism.  Moreover, when we apply the dslash operator, since the length of each dimension is small, we are much more susceptible to warp divergence issues, e.g., threads on the boundary load from the ghost whereas as threads in the bulk load from the main field.  

A matrix powers approach would rectify this:
1. Do all the communication up front for `s` applications of the operator
2. Do `s` applications of the operator

While this approach introduces redundancy since adjacent nodes will be repeating computation, the advantages are that
* Message aggregation allows for the saturation of interconnect bandwidth
* The operator kernel simply acts on an extended field, with the halo regions inlined into it.  Thus there are no warp divergence issues

In the coarse limit, there will be no overhead from the redundant computation owing to the increased efficiency, and any additional overhead will be covered by available GPU resources.

This is clearly an optimization for further down the road....

**Comment by maddyscientist** [comment]
Closing this issue, since we have CA-GCR which provides the functionality we need, so there is no desperate need for CA-GMRES at this time.  Nice to have, but no immediate advantage versus CA-GCR.
